kbs/docs/config.md
# KBS Configuration File

The Confidential Containers KBS properties can be configured through a
TOML-formatted configuration file.

>NOTE: Additional formats such as YAML and JSON are supported. Other formats
>supported by the `config` crate may be supported as well. This document uses
>TOML in the configuration examples.

The location of the configuration file is passed to the KBS binary using the
`-c` or `--config-file` command line option, or using the `KBS_CONFIG_FILE`
environment variable.

## Configurable Properties

The following sections list the KBS properties which can be set through the
configuration file.

### HTTP Server Configuration

The following properties can be set under the `[http_server]` section.

| Property                 | Type         | Description                                                                                                | Required | Default              |
|--------------------------|--------------|------------------------------------------------------------------------------------------------------------|----------|----------------------|
| `sockets`                | String array | One or more sockets to listen on.                                                                          | No       | `["127.0.0.1:8080"]` |
| `insecure_http`          | Boolean      | Don't use TLS for the KBS HTTP endpoint.                                                                   | No       | `false`              |
| `private_key`            | String       | Path to a private key file to be used for HTTPS.                                                           | No       | None                 |
| `certificate`            | String       | Path to a certificate file to be used for HTTPS.                                                           | No       | None                 |

### Attestation Token Configuration

Attestation Token configuration controls attestation token verifications. This
is important when a resource retrievement is handled by KBS. Usually an attestation
token will be together with the request, and KBS will first verify the token.

The following properties can be set under the `[attestation_token]` section.

| Property                   | Type         | Description                                                                                                                                               | Default |
|----------------------------|--------------|----------------------------------------------------------------------------------------------------------------------------------------------------------|----------|
| `trusted_jwk_sets` | String Array      | Valid Url (`file://` or `https://`) pointing to trusted JWKSets (local or OpenID) for Attestation Tokens trustworthy verification                                                                                             | Empty       |
| `trusted_certs_paths` | String Array | Trusted Certificates file (PEM format) for Attestation Tokens trustworthy verification | Empty       |
| `extra_teekey_paths` | String Array | User defined paths to the tee public key in the JWT body  | Empty       |
| `insecure_key` | Boolean | Whether to check the trustworthy of the JWK inside JWT. See comments. | `false`      |

Each JWT contains a TEE Public Key. Users can use the `extra_teekey_paths` field to additionally specify the path of this Key in the JWT.
Example of `extra_teekey_paths` is `/attester_runtime_data/tee-pubkey` which refers to the key
`attester_runtime_data.tee-pubkey` inside the JWT body claims. By default CoCo AS Token and Intel TA
Token TEE Public Key paths are supported.

For Attestation Services like CoCo-AS, the public key to verify the JWT will be given
in the token's `jwk` field (with or without the public key cert chain `x5c`).

- If `insecure_key` is set to `true`, KBS will ignore to verify the trustworthy of the `jwk`.
- If `insecure_key` is set to `false`, KBS will look up its `trusted_certs_paths` and the `x5c`
field to verify the trustworthy of the `jwk`.

For Attestation Services like Intel TA, there will only be a `kid` field inside the JWT.
The `kid` field is used to look up the trusted jwk configured by KBS via `trusted_jwk_sets` to
verify the integrity and trustworthy of the JWT.

### Attestation Configuration

Attestation configuration defines the attestation service that KBS' RCAR protocol
will leverage.

The following properties can be set under the `[attestation_service]` section.

Concrete attestation service can be set via `type` field. Supported attestation
services are
- `coco_as_builtin`: CoCo AS that built inside KBS binary
- `coco_as_grpc`: CoCo AS service running remotely
- `intel_ta`: Intel&reg; Trust Authority

Due to different `type` field, properties are different.

#### Built-In CoCo AS

When `type` is set to `coco_as_builtin`, the following properties can be set.

>Built-In CoCo AS is available only when one or more of the following features are enabled:
>`coco-as-builtin`, `coco-as-builtin-no-verifier`

| Property                   | Type                        | Description                                          | Default |
|----------------------------|-----------------------------|-----------------------------------------------------|----------|
| `timeout`            | Integer                      | The maximum time (in minutes) between RCAR handshake's `auth` and `attest` requests             |  5       |
| `work_dir`                 | String                      | The location for Attestation Service to store data. |  First try from env `AS_WORK_DIR`. If no this env, then use `/opt/confidential-containers/attestation-service`       |
| `policy_engine`            | String                      | Policy engine type. Valid values: `opa`             |  `opa`       |
| `rvps_config`              | [RVPSConfiguration][2]      | RVPS configuration                                  |  See [RVPSConfiguration][2]       |
| `attestation_token_broker` | String                      | Type of the attestation result token broker.        |  `Simple`       |
| `attestation_token_config` | [AttestationTokenConfig][1] | Attestation result token configuration.             |  See [AttestationTokenConfig][1]       |

[1]: #attestationtokenconfig
[2]: #rvps-configuration


##### AttestationTokenConfig

| Property       | Type                    | Description                                          | Default |
|----------------|-------------------------|------------------------------------------------------|----------|
| `duration_min` | Integer                 | Duration of the attestation result token in minutes. | 5       |
| `issuer_name`  | String                  | Issure name of the attestation result token.         | `CoCo-Attestation-Service`       |
| `signer`       | [TokenSignerConfig][1]  | Signing material of the attestation result token.    | None       |

[1]: #tokensignerconfig

##### TokenSignerConfig

This section is **optional**. When omitted, an ephemeral RSA key pair is generated and used. 

| Property       | Type    | Description                                              | Required |
|----------------|---------|----------------------------------------------------------|----------|
| `key_path`     | String  | RSA Key Pair file (PEM format) path.                     | Yes      |
| `cert_url`     | String  | RSA Public Key certificate chain (PEM format) URL.       | No       |
| `cert_path`    | String  | RSA Public Key certificate chain (PEM format) file path. | No       |

##### RVPS Configuration

| Property       | Type                    | Description                                           | Default |
|----------------|-------------------------|------------------------------------------------------|---------|
| `remote_addr`  | String                  | Remote RVPS' address. If this is specified, will use a remote RVPS. Or a local RVPS will be configured with `store_type` and `store_config`| Empty       |
| `store_type`   | String                  | Used if `remote_addr` is not set. The underlying storage type of RVPS.                                                                     | `LocalFs`       |
| `store_config` | JSON Map                | Used if `remote_addr` is not set. The optional configurations to the underlying storage.                                                   | Empty       |

Different `store_type` will have different `store_config` items.
See the details of `store_config` in [concrete implementations of storages](../../rvps/src/store/).

#### gRPC CoCo AS

When `type` is set to `coco_as_grpc`, KBS will try to connect a remote CoCo AS for
attestation. The following properties can be set.

>gRPC CoCo AS is available only when `coco-as-grpc` feature is enabled.

| Property                   | Type                        | Description                                          | Default |
|----------------------------|-----------------------------|-----------------------------------------------------|----------|
| `timeout`            | Integer                      | The maximum time (in minutes) between RCAR handshake's `auth` and `attest` requests             |  5       |
| `as_addr`                 | String                      | The URL of the remote CoCoAS |  `http://127.0.0.1:50004`       |
| `pool_size`   | Integer         | The connections between KBS and CoCoAS are maintained in a conenction pool. This property determines the max size of the pool                      | `100`             |

#### Intel&reg; TA

When `type` is set to `intel_ta`, KBS will try to connect a remote Intel TA service for
attestation. The following properties can be set.

> Intel Trust Authority AS is available only when the `intel-trust-authority-as` feature is enabled.

| Property                 | Type         | Description                                                                              | Required | Default |
|--------------------------|--------------|------------------------------------------------------------------------------------------|----------|---------|
| `timeout`                | Integer      | The maximum time (in minutes) between RCAR handshake's `auth` and `attest` requests      | No       | 5       |
| `base_url`               | String       | Intel Trust Authority API URL.                                                           | Yes      | -       |
| `api_key`                | String       | Intel Trust Authority API key.                                                           | Yes      | -       |
| `certs_file`             | String       | URL to an Intel Trust Authority portal or path to JWKS file used for token verification. | Yes      | -       |
| `allow_unmatched_policy` | Boolean      | If set and `policy_ids` specified, unset the `request.policy_must_match` setting         | No       | false   |
| `policy_ids`             | String array | List of one or more quoted and comma-separated policy IDs.                               | No       | `[]`    |

Detailed [documentation](https://docs.trustauthority.intel.com).

### Admin API Configuration

The following properties can be set under the `[admin]` section.

| Property                 | Type         | Description                                                                                                | Required | Default              |
|--------------------------|--------------|------------------------------------------------------------------------------------------------------------|----------|----------------------|
| `auth_public_key`        | String       | Path to the public key used to authenticate the admin APIs                                                 | No       | None                 |
| `insecure_api`           | Boolean      | Whether KBS will not verify the public key when called admin APIs                                          | No       | `false`              |

### Policy Engine Configuration

The following properties can be set under the `[policy_engine]` section.

This section is **optional**. When omitted, a default configuration is used.

| Property                 | Type    | Description                                                                                                | Required                | Default                                        |
|--------------------------|---------|------------------------------------------------------------------------------------------------------------|-------------------------|------------------------------------------------|
| `policy_path`            | String  | Path to a file containing a policy for evaluating whether the TCB status has access to specific resources. | No                      | `/opa/confidential-containers/kbs/policy.rego` |

### Plugins Configuration

KBS supports different kinds of plugins, and they can be enabled via add corresponding configs.

Multiple `[[plugins]]` sections are allowed at the same time for different plugins.
Concrete attestation service can be set via `name` field.

#### Resource Configuration

The `name` field is `resource` to enable this plugin.

Resource plugin allows user with proper attestation token to access storage that KBS keeps.
This is also called "Repository" in old versions. The properties to be configured are listed.

| Property | Type   | Description                                                     | Required | Default   |
|----------|--------|-----------------------------------------------------------------|----------|-----------|
| `type`   | String | The resource repository type. Valid values: `LocalFs`, `Aliyun` | Yes      | `LocalFs` |

**`LocalFs` Properties**

| Property   | Type   | Description                     | Required | Default                                             |
|------------|--------|---------------------------------|----------|-----------------------------------------------------|
| `dir_path` | String | Path to a repository directory. | No       | `/opt/confidential-containers/kbs/repository`       |

**`Aliyun` Properties**

| Property          | Type   | Description                       | Required | Example                                             |
|-------------------|--------|-----------------------------------|----------|-----------------------------------------------------|
| `client_key`      | String | The KMS instance's AAP client key | Yes      | `{"KeyId": "KA..", "PrivateKeyData": "MIIJqwI..."}` |
| `kms_instance_id` | String | The KMS instance id               | Yes      | `kst-shh668f7...`                                   |
| `password`        | String | AAP client key password           | Yes      | `8f9989c18d27...`                                   |
| `cert_pem`        | String | CA cert for the KMS instance      | Yes      | `-----BEGIN CERTIFICATE----- ...`                   |

## Configuration Examples

Using a built-in CoCo AS:

```toml
[http_server]
sockets = ["0.0.0.0:8080"]
insecure_http = true

[admin]
insecure_api = true

[attestation_token]

[attestation_service]
type = "coco_as_builtin"
work_dir = "/opt/confidential-containers/attestation-service"
policy_engine = "opa"

    [attestation_service.attestation_token_broker]
    type = "Ear"
    duration_min = 5

    [attestation_service.rvps_config]
    store_type = "LocalFs"
    remote_addr = ""

[[plugins]]
name = "resource"
type = "LocalFs"
dir_path = "/opt/confidential-containers/kbs/repository"
```

Using a remote CoCo AS:

```toml
[http_server]
insecure_http = true

[admin]
insecure_api = true

[attestation_service]
type = "coco_as_grpc"
as_addr = "http://127.0.0.1:50004"

[[plugins]]
name = "resource"
type = "LocalFs"
dir_path = "/opt/confidential-containers/kbs/repository"
```

Running with Intel Trust Authority attestation service:

```toml
[http_server]
sockets = ["0.0.0.0:8080"]
private_key = "/etc/kbs-private.key"
certificate = "/etc/kbs-cert.pem"
insecure_http = false

[attestation_token]
trusted_jwk_sets = ["https://portal.trustauthority.intel.com"]

[attestation_token]

[attestation_service]
type = "intel_ta"
base_url = "https://api.trustauthority.intel.com"
api_key = "tBfd5kKX2x9ahbodKV1..."
certs_file = "https://portal.trustauthority.intel.com"
allow_unmatched_policy = true

[admin]
auth_public_key = "/etc/kbs-admin.pub"
insecure_api = false

[policy_engine]
policy_path = "/etc/kbs-policy.rego"

[[plugins]]
name = "resource"
type = "LocalFs"
dir_path = "/opt/confidential-containers/kbs/repository"
```

Distributing resources in Passport mode:

```toml
[http_server]
sockets = ["127.0.0.1:50002"]
insecure_http = true

[admin]
auth_public_key = "./work/kbs.pem"

[attestation_token]
trusted_certs_paths = ["./work/ca-cert.pem"]
insecure_key = false

[policy_engine]

kbs/src/config.rs
// Copyright (c) 2023 by Alibaba.
// Licensed under the Apache License, Version 2.0, see LICENSE for details.
// SPDX-License-Identifier: Apache-2.0

use crate::admin::config::{AdminConfig, DEFAULT_INSECURE_API};
use crate::plugins::PluginsConfig;
use crate::policy_engine::PolicyEngineConfig;
use crate::token::AttestationTokenVerifierConfig;
use anyhow::anyhow;
use clap::Parser;
use config::{Config, File};
use serde::Deserialize;
use std::net::SocketAddr;
use std::path::{Path, PathBuf};

const DEFAULT_INSECURE_HTTP: bool = false;
const DEFAULT_SOCKET: &str = "127.0.0.1:8080";

#[derive(Clone, Debug, Deserialize, PartialEq)]
pub struct HttpServerConfig {
    /// Socket addresses (IP:port) to listen on, e.g. 127.0.0.1:8080.
    pub sockets: Vec<SocketAddr>,

    /// HTTPS private key.
    pub private_key: Option<PathBuf>,

    /// HTTPS Certificate.
    pub certificate: Option<PathBuf>,

    /// Insecure HTTP.
    /// WARNING: Using this option makes the HTTP connection insecure.
    pub insecure_http: bool,
}

impl Default for HttpServerConfig {
    fn default() -> Self {
        Self {
            sockets: vec![DEFAULT_SOCKET.parse().expect("unexpected parse error")],
            private_key: None,
            certificate: None,
            insecure_http: DEFAULT_INSECURE_HTTP,
        }
    }
}

/// Contains all configurable KBS properties.
#[derive(Debug, Clone, Deserialize, PartialEq)]
pub struct KbsConfig {
    /// Attestation token result broker config.
    #[serde(default)]
    pub attestation_token: AttestationTokenVerifierConfig,

    /// Configuration for the Attestation Service.
    #[cfg(feature = "as")]
    #[serde(default)]
    pub attestation_service: crate::attestation::config::AttestationConfig,

    /// Configuration for the KBS Http Server
    pub http_server: HttpServerConfig,

    /// Configuration for the KBS admin API
    pub admin: AdminConfig,

    /// Policy engine configuration used for evaluating whether the TCB status has access to
    /// specific resources.
    #[serde(default)]
    pub policy_engine: PolicyEngineConfig,

    #[serde(default)]
    pub plugins: Vec<PluginsConfig>,
}

impl TryFrom<&Path> for KbsConfig {
    type Error = anyhow::Error;

    /// Load `Config` from a configuration file. Supported formats are all formats supported by the
    /// `config` crate. See `KbsConfig` for schema information.
    fn try_from(config_path: &Path) -> Result<Self, Self::Error> {
        let c = Config::builder()
            .set_default("admin.insecure_api", DEFAULT_INSECURE_API)?
            .set_default("http_server.insecure_http", DEFAULT_INSECURE_HTTP)?
            .set_default("http_server.sockets", vec![DEFAULT_SOCKET])?
            .set_default("attestation_service.policy_ids", Vec::<&str>::new())?
            .add_source(File::with_name(config_path.to_str().unwrap()))
            .build()?;

        c.try_deserialize()
            .map_err(|e| anyhow!("invalid config: {}", e.to_string()))
    }
}

/// KBS command-line arguments.
#[derive(Debug, Parser)]
#[command(author, version, about, long_about = None)]
pub struct Cli {
    /// Path to a KBS config file. Supported formats: TOML, YAML, JSON and possibly other formats
    /// supported by the `config` crate.
    #[arg(short, long, env = "KBS_CONFIG_FILE")]
    pub config_file: String,
}

#[cfg(test)]
mod tests {
    use std::path::{Path, PathBuf};

    use crate::{
        admin::config::AdminConfig,
        config::{HttpServerConfig, DEFAULT_INSECURE_API, DEFAULT_INSECURE_HTTP, DEFAULT_SOCKET},
        plugins::{
            implementations::{
                resource::local_fs::LocalFsRepoDesc, RepositoryConfig, SampleConfig,
            },
            PluginsConfig,
        },
        policy_engine::{PolicyEngineConfig, DEFAULT_POLICY_PATH},
        token::{AttestationTokenVerifierConfig, AttestationTokenVerifierType},
    };

    use super::KbsConfig;

    #[cfg(feature = "coco-as-builtin")]
    use attestation_service::{
        rvps::{RvpsConfig, DEFAULT_STORAGE_TYPE},
        token::{simple, AttestationTokenConfig, COCO_AS_ISSUER_NAME, DEFAULT_TOKEN_DURATION},
    };

    use rstest::rstest;
    use serde_json::json;

    #[rstest]
    #[case("test_data/configs/coco-as-grpc-1.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            attestation_token_type: AttestationTokenVerifierType::Ear,
            trusted_certs_paths: vec!["/etc/ca".into(), "/etc/ca2".into()],
            insecure_key: false,
            trusted_jwk_sets: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "coco-as-grpc")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASGrpc(
                    crate::attestation::coco::grpc::GrpcConfig {
                        as_addr: "http://127.0.0.1:50001".into(),
                        pool_size: 100,
                    },
                ),
            timeout: 600,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: Some("/etc/kbs-private.key".into()),
            certificate: Some("/etc/kbs-cert.pem".into()),
            insecure_http: false,
        },
        admin: AdminConfig {
            auth_public_key: Some(PathBuf::from("/etc/kbs-admin.pub")),
            insecure_api: false,
        },
        policy_engine: PolicyEngineConfig {
            policy_path: PathBuf::from("/etc/kbs-policy.rego"),
        },
        plugins: vec![PluginsConfig::Sample(SampleConfig {
            item: "value1".into(),
        }),
        PluginsConfig::ResourceStorage(RepositoryConfig::LocalFs(
            LocalFsRepoDesc {
                dir_path: "/tmp/kbs-resource".into(),
            },
        ))],
    })]
    #[case("test_data/configs/coco-as-builtin-1.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            attestation_token_type: AttestationTokenVerifierType::Ear,
            trusted_certs_paths: vec![],
            insecure_key: false,
            trusted_jwk_sets: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "coco-as-builtin")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASBuiltIn(
                    attestation_service::config::Config {
                        work_dir: "/opt/coco/attestation-service".into(),
                        rvps_config: RvpsConfig {
                            remote_addr: "http://127.0.0.1:50003".into(),
                            store_type: DEFAULT_STORAGE_TYPE.into(),
                            store_config: json!({}),
                        },
                        attestation_token_broker: AttestationTokenConfig::Simple(simple::Configuration {
                            duration_min: DEFAULT_TOKEN_DURATION,
                            issuer_name: COCO_AS_ISSUER_NAME.into(),
                            signer: None,
                            ..Default::default()
                        }),
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec![DEFAULT_SOCKET.parse().unwrap()],
            private_key: None,
            certificate: None,
            insecure_http: DEFAULT_INSECURE_HTTP,
        },
        admin: AdminConfig {
            auth_public_key: None,
            insecure_api: DEFAULT_INSECURE_API,
        },
        policy_engine: PolicyEngineConfig {
            policy_path: DEFAULT_POLICY_PATH.into(),
        },
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/intel-ta-1.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            attestation_token_type: AttestationTokenVerifierType::Ear,
            trusted_jwk_sets: vec!["/etc/ca".into(), "/etc/ca2".into()],
            insecure_key: false,
            trusted_certs_paths: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "intel-trust-authority-as")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::IntelTA(
                    crate::attestation::intel_trust_authority::IntelTrustAuthorityConfig {
                        base_url: "example.io".into(),
                        api_key: "this-is-a-key".into(),
                        certs_file: "file:///etc/ita-cert.pem".into(),
                        allow_unmatched_policy: Some(true),
                        policy_ids: vec![],
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: Some("/etc/kbs-private.key".into()),
            certificate: Some("/etc/kbs-cert.pem".into()),
            insecure_http: false,
        },
        admin: AdminConfig {
            auth_public_key: Some(PathBuf::from("/etc/kbs-admin.pub")),
            insecure_api: false,
        },
        policy_engine: PolicyEngineConfig {
            policy_path: PathBuf::from("/etc/kbs-policy.rego"),
        },
        plugins: vec![PluginsConfig::Sample(SampleConfig {
            item: "value1".into(),
        }),
        PluginsConfig::ResourceStorage(RepositoryConfig::LocalFs(
            LocalFsRepoDesc {
                dir_path: "/tmp/kbs-resource".into(),
            },
        ))],
    })]
    #[case("test_data/configs/coco-as-grpc-2.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            ..Default::default()
        },
        #[cfg(feature = "coco-as-grpc")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASGrpc(
                    crate::attestation::coco::grpc::GrpcConfig {
                        as_addr: "http://as:50004".into(),
                        pool_size: crate::attestation::coco::grpc::DEFAULT_POOL_SIZE,
                    },
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: None,
            certificate: None,
            insecure_http: true,
        },
        admin: AdminConfig {
            auth_public_key: Some(PathBuf::from("/opt/confidential-containers/kbs/user-keys/public.pub")),
            insecure_api: DEFAULT_INSECURE_API,
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/coco-as-builtin-2.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            attestation_token_type: AttestationTokenVerifierType::Ear,
            trusted_certs_paths: vec![],
            insecure_key: false,
            trusted_jwk_sets: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "coco-as-builtin")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASBuiltIn(
                    attestation_service::config::Config {
                        work_dir: "/opt/confidential-containers/attestation-service".into(),
                        rvps_config: RvpsConfig {
                            remote_addr: "".into(),
                            store_type: "LocalFs".into(),
                            store_config: json!({}),
                        },
                        attestation_token_broker: AttestationTokenConfig::Simple(simple::Configuration{
                            duration_min: 5,
                            ..Default::default()
                        }),
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: None,
            certificate: None,
            insecure_http: true,
        },
        admin: AdminConfig {
            auth_public_key: Some("/kbs/kbs.pem".into()),
            insecure_api: DEFAULT_INSECURE_API,
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/intel-ta-2.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            attestation_token_type: AttestationTokenVerifierType::Ear,
            trusted_jwk_sets: vec!["https://portal.trustauthority.intel.com".into()],
            insecure_key: false,
            trusted_certs_paths: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "intel-trust-authority-as")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::IntelTA(
                    crate::attestation::intel_trust_authority::IntelTrustAuthorityConfig {
                        base_url: "https://api.trustauthority.intel.com".into(),
                        api_key: "tBfd5kKX2x9ahbodKV1...".into(),
                        certs_file: "https://portal.trustauthority.intel.com".into(),
                        allow_unmatched_policy: None,
                        policy_ids: vec![],
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: None,
            certificate: None,
            insecure_http: true,
        },
        admin: AdminConfig {
            auth_public_key: Some("/kbs/kbs.pem".into()),
            insecure_api: DEFAULT_INSECURE_API,
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/coco-as-grpc-3.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            ..Default::default()
        },
        #[cfg(feature = "coco-as-grpc")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASGrpc(
                    crate::attestation::coco::grpc::GrpcConfig {
                        as_addr: "http://127.0.0.1:50004".into(),
                        pool_size: 100,
                    },
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            insecure_http: true,
            ..Default::default()
        },
        admin: AdminConfig {
            insecure_api: true,
            ..Default::default()
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/intel-ta-3.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            attestation_token_type: AttestationTokenVerifierType::Ear,
            trusted_jwk_sets: vec!["https://portal.trustauthority.intel.com".into()],
            insecure_key: false,
            trusted_certs_paths: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "intel-trust-authority-as")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::IntelTA(
                    crate::attestation::intel_trust_authority::IntelTrustAuthorityConfig {
                        base_url: "https://api.trustauthority.intel.com".into(),
                        api_key: "tBfd5kKX2x9ahbodKV1...".into(),
                        certs_file: "https://portal.trustauthority.intel.com".into(),
                        allow_unmatched_policy: None,
                        policy_ids: vec![],
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            insecure_http: true,
            ..Default::default()
        },
        admin: AdminConfig {
            insecure_api: true,
            ..Default::default()
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/coco-as-builtin-3.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            attestation_token_type: AttestationTokenVerifierType::Ear,
            trusted_certs_paths: vec![],
            insecure_key: false,
            trusted_jwk_sets: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "coco-as-builtin")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASBuiltIn(
                    attestation_service::config::Config {
                        work_dir: "/opt/confidential-containers/attestation-service".into(),
                        rvps_config: RvpsConfig {
                            remote_addr: "".into(),
                            store_type: "LocalFs".into(),
                            ..Default::default()
                        },
                        attestation_token_broker: AttestationTokenConfig::Simple(simple::Configuration {
                            duration_min: 5,
                            ..Default::default()
                        }),
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            insecure_http: true,
            ..Default::default()
        },
        admin: AdminConfig {
            insecure_api: true,
            ..Default::default()
        },
        policy_engine: PolicyEngineConfig {
            policy_path: "/opa/confidential-containers/kbs/policy.rego".into(),
        },
        plugins: vec![
        PluginsConfig::ResourceStorage(RepositoryConfig::LocalFs(
            LocalFsRepoDesc {
                dir_path: "/opt/confidential-containers/kbs/repository".into(),
            },
        ))],
    })]
    fn read_config(#[case] config_path: &str, #[case] expected: KbsConfig) {
        let config = KbsConfig::try_from(Path::new(config_path)).unwrap();
        assert_eq!(config, expected, "case {config_path}");
    }
}
